src/Motor.Extensions.Hosting.CloudEvents/MotorCloudEvent.cs

```diff
@@ -0,0 +1,114 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Mime;
+using CloudNative.CloudEvents;
+
+namespace Motor.Extensions.Hosting.CloudEvents
+{
+    public static class MotorCloudEventInfo
+    {
+        public static CloudEventsSpecVersion SpecVersion => CloudEventsSpecVersion.V1_0;
+
+        public static readonly IEnumerable<CloudEventAttribute> RequiredAttributes = SpecVersion.RequiredAttributes.Concat(new[]{
+            SpecVersion.TimeAttribute, SpecVersion.DataContentTypeAttribute
+        });
+    }
+
+    public class MotorCloudEvent<TData> where TData : class
+    {
+        private readonly IApplicationNameService _applicationNameService;
+
+        public MotorCloudEvent(
+            IApplicationNameService applicationNameService,
+            TData data,
+            Uri source,
+            params KeyValuePair<CloudEventAttribute, object>[] extensions) : this(applicationNameService, data, source,
+            null, null, extensions)
+        {
+        }
+
+        public MotorCloudEvent(
+            IApplicationNameService applicationNameService,
+            TData data,
+            Uri source,
+            string? id,
+            DateTimeOffset? time,
+            params KeyValuePair<CloudEventAttribute, object>[] extensions)
+        {
+            BaseEvent = new CloudEvent(CloudEventsSpecVersion.Default);
+            foreach (var (key, value) in extensions) BaseEvent[key] = value;
+            BaseEvent.Id = id ?? Guid.NewGuid().ToString();
+            BaseEvent.Type = typeof(TData).Name;
+            BaseEvent.Source = source;
+            BaseEvent.Time = time ?? DateTimeOffset.UtcNow;
+            BaseEvent.DataContentType = new ContentType().ToString();
+
+            _applicationNameService = applicationNameService;
+            TypedData = data;
+        }
+
+        public TData TypedData
+        {
+            get => (TData?)BaseEvent.Data!;
+            set => BaseEvent.Data = value;
+        }
+
+        public object Data => BaseEvent.Data!;
+        public string Id => BaseEvent.Id!;
+        public string Type => BaseEvent.Type!;
+        public DateTimeOffset Time => BaseEvent.Time!.Value;
+        public Uri Source => BaseEvent.Source!;
+
+        public object? this[CloudEventAttribute attribute]
```

Comment:
Codacy is not happy about using a custom class for indexing. However, we built this class as a wrapper of `CloudEvent` and thus used the same syntactical features.